libraries/AP_RangeFinder/AP_RangeFinder_BLPing.cpp

```diff
@@ -0,0 +1,286 @@
+#define BLPING_MSGID_GET_DEVICE_ID          1201
+#define BLDPIN_MSGID_DISTANCE_SIMPLE        1211
+#define BLPING_MSGID_CONTINUOUS_START       1400
+
+// Protocol implemented by this sensor can be found here: https://github.com/bluerobotics/ping-protocol
+//
+// Byte     Type        Name            Description
+// --------------------------------------------------------------------------------------------------------------
+// 0        uint8_t     start1          'B'
+// 1        uint8_t     start2          'R'
+// 2-3      uint16_t    payload_length  number of bytes in payload (low byte, high byte)
+// 4-5      uint16_t    message id      message id (low byte, high byte)
+// 6        uint8_t     src_device_id   id of device sending the message
+// 7        uint8_t     dst_device_id   id of device of the intended recipient
+// 8-n      uint8_t[]   payload         message payload
+// (n+1)-(n+2)  uint16_t    checksum    the sum of all the non-checksum bytes in the message (low byte, high byte)
+
+/* 
+   The constructor also initialises the rangefinder. Note that this
+   constructor is not called until detect() returns true, so we
+   already know that we should setup the rangefinder
+*/
+AP_RangeFinder_BLPing::AP_RangeFinder_BLPing(RangeFinder::RangeFinder_State &_state,
+                                                             AP_RangeFinder_Params &_params,
+                                                             AP_SerialManager &serial_manager,
+                                                             uint8_t serial_instance) :
+    AP_RangeFinder_Backend(_state, _params)
+{
+    uart = serial_manager.find_serial(AP_SerialManager::SerialProtocol_Rangefinder, serial_instance);
+    if (uart != nullptr) {
+        uart->begin(serial_manager.find_baudrate(AP_SerialManager::SerialProtocol_Rangefinder, serial_instance));
+    }
+}
+
+// detect if a serial port has been setup to accept rangefinder input
+bool AP_RangeFinder_BLPing::detect(AP_SerialManager &serial_manager, uint8_t serial_instance)
+{
+    return serial_manager.find_serial(AP_SerialManager::SerialProtocol_Rangefinder, serial_instance) != nullptr;
+}
+
+/*
+   update the state of the sensor
+*/
+void AP_RangeFinder_BLPing::update(void)
+{
+    uint32_t now = AP_HAL::millis();
+
+    if (get_reading(state.distance_cm)) {
+        // update range_valid state based on distance measured
+        state.last_reading_ms = AP_HAL::millis();
+        update_status();
+    } else if (now - state.last_reading_ms > BLPING_TIMEOUT_MS) {
+        set_status(RangeFinder::RangeFinder_NoData);
+
+        // initialise sensor if no distances recently
+        if (now - last_init_ms > BLPING_INIT_RATE_MS) {
+            last_init_ms = now;
+            init_sensor();
+        }
+    }
+}
+
+// initialise sensor
+void AP_RangeFinder_BLPing::init_sensor()
+{
+    // request distance from sensor
+    send_message(BLDPIN_MSGID_DISTANCE_SIMPLE, nullptr, 0);
```

Comment:
perhaps although the Lightware serial range finder is similar.  We always send it a "d" if we want a new reading.  I would perhaps also prefer that the sensor just sends data as quickly as it can which result in a faster update rate.